### DISCUSSION TITLE
c++ brace initialisation

### DISCUSSION THREAD

**Discussion by yyueyangg** [community/community#57414] [Programming Help]
### Body

Hello, I am using Mac OS Apple clang version 12.0.0 (clang-1200.0.32.29) in vs code to learn c++. I stumbled upon brace initialisation on a book and would like to try it out. However, an error message (expected ';' at the end of declaration) keeps popping up. I tried to change my c++ version in c_cpp_properties.json to c++11, c++17 and even c++23 as well but the error message is still there. Does anyone have any advice on how I can resolve this issue? Thank you guys .

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by yyueyangg** [comment]
I have also disabled any relevant extensions in case those are the things causing the error messages to pop up, but it is futile

**Reply by jge162** [reply]
Do you have a sample of code to see? Thanks.

**Reply by yyueyangg** [reply]
This is the code 

# include <iostream>

using namespace std;

int main()
{
    int s{200};
    cout << s << endl;
}

**Reply by jge162** [reply]
Looks fine. Issue is your settings then.

**Reply by yyueyangg** [reply]
yes. I think so too. I have gone to c_cpp_properties.json to change my c++ ver to 11, 17, 23, 20 and tried with all the versions. None works. I also disbked any kinds of extensions that may have generated the error message. Do u have any idea what is wrong in the settings?

**Comment by davea42** [comment]
Your example code is missing the required line

`#include <iostream>`

before the function declaration.
No doubt you figured that out weeks ago...

**Reply by GhostOf0days** [reply]
also, return 0; at end of main. It's int main().

**Reply by yyueyangg** [reply]
the include line was typed out when I tried the code but the error still stays 😩

**Reply by hardy-ethan** [reply]
> also, return 0; at end of main. It's int main().

actually, int main automatically returns 0 if omitted in (at least modern) C++

**Reply by davea42** [reply]
ANSI C 1989/ISO C 1990  both specified that main() returns 0 if no return value is specified
in the code.   So every standard version of C++ does too.  Before the C standard existed K&R C
did not specify what happens on running off the end of main() even though every code example
in K&R that I found fails to provide a return from main()!  So I suspect K&R C provided that special-case
return too, though I can no longer remember for sure.  K&R does not mention main() as having
a special aspect to the return statement.

(I just looked through "The C Programming Language" by Kerrnighan and Ritchie, copyright 1978, to refresh my recollection of K&R C.)

**Reply by hardy-ethan** [reply]
> ANSI C 1989/ISO C 1990 both specified that main() returns 0 if no return value is specified in the code. So every standard version of C++ does too. Before the C standard existed K&R C did not specify what happens on running off the end of main() even though every code example in K&R that I found fails to provide a return from main()! So I suspect K&R C provided that special-case return too, though I can no longer remember for sure. K&R does not mention main() as having a special aspect to the return statement.
> 
> (I just looked through "The C Programming Language" by Kerrnighan and Ritchie, copyright 1978, to refresh my recollection of K&R C.)

nice!

**Comment by davea42** [comment]
Putting on my language-lawyer hat, I observe that in both C and C++
if a main() function  does not have a return statement at the end
the compiler adds a return 0; for you.

The compiler(s) **only** add such if the function is named main!

Of course return 0; is good because it reveals the programmer intent
so GhostOf0days is...not wrong.

**Comment by HammoucheSonia** [comment]
It seems like you are facing an issue with brace initialization in C++ on your VS Code setup. The problem might not be related to the C++ version settings in `c_cpp_properties.json` but could be due to incorrect code syntax or other configuration issues. Here are some steps you can take to resolve the problem:

1. **Code Syntax:** Double-check the brace initialization syntax in your code. Make sure you are using the correct syntax for brace initialization in C++. For example:
   
   ```cpp
   int myVar = {42};
   ```

   If you are initializing a class or struct, ensure that the appropriate constructor exists.

2. **Compiler Settings:** Ensure that you have set the correct C++ version in the VS Code settings. To do this, open the `settings.json` file for your workspace (or user settings if you want it to apply globally) and add the following:

   ```json
   {
       "C_Cpp.default.cppStandard": "c++17"
   }
   ```

   Replace `"c++17"` with the desired C++ version (e.g., `"c++11"` or `"c++23"`).

3. **Check for Typos:** Sometimes, errors can be due to typos or missing semicolons in the code. Double-check your code for any mistakes.

4. **Compiler Extensions:** Ensure that any compiler extensions or plugins you might be using are compatible with the C++ version you want to use.

5. **VS Code Extension:** Consider installing the "C/C++" extension by Microsoft in VS Code, which provides rich language support for C++ and can help with proper syntax highlighting and error checking.

6. **Restart VS Code:** After making changes to the settings, it's a good idea to restart VS Code to ensure the changes take effect.

7. **Makefile or Build System:** If you are using a custom build system or Makefile, ensure that it is correctly configured for the C++ version you want to use.

If you have tried all of the above and are still facing issues, please double-check your code and provide more information about the specific error and the code causing it so that I can assist you further.

**Reply by yyueyangg** [reply]
# include <iostream>

using namespace std;

int main()
{
    int a = {200};
    cout << a << endl;
    int b{100};
    cout << b << endl;
}

this is a version of my code, a can be printed out but not b. 
I have also tried changing the c++ version in the compiler settings, checked for typos, downloaded extensions from Microsoft and disabled any extensions that may affect the code and restarted vs code. all these still doesn't work. may I know if there are any other solutions?

**Reply by yyueyangg** [reply]
the include line was written together with my code in vs code but I'm not sure why when I copied and pasted it here it disappears, but no worries I typed it together with the rest of the code when im coding for the brace initialisation so it shouldn't be an include line error

**Comment by davea42** [comment]
So I fired up a Mac Mini with the apple command line tools installed,
and with the include <iostream> that you left out above I too get
a n error about a missing ';'.
 c++ --version shows
words like clang version 14.0.3

int s{200};    gets an error about semicolon.
On spending a few minutes with the c++ 1998 (!) standard
I seem to find that, in section 8.5, Initializers
nothing allows this version as legal C++. But
I am not very confident of my reading of the
C++ standard and well, it is ...an ancient... version.

So then I looked up Aggregate Initialization on the web
and it appears (?) your chosen form 
of initialization was made legal  in C++11.

int s(200);   works fine and on running the program prints 200.
int s =(200);   works fine and on running the program prints 200.
int s=200;    works fine and on running the program prints 200.
int s= {200} works fine and on running the program prints 200.

In contrast, g++ (on linux) compiles all versions and all produce the same output,
with s having value 200.

**Reply by yyueyangg** [reply]
may I have a version of your code? or is it possible if u take a look at my code in the preceding comment and suggest to me what is the difference between your code and mine? int s{200} doesn't work for me

**Comment by davea42** [comment]
Yes, on Mac  case A:
int s{200};  fails in compilation.
It appears we are in agreement about Macos failing that one initialization.

The other four versions I show above  compile everywhere (I claim).

Here is code compiling each on Linux.
 If you do the same on Macos case A
will fail and the others will compile ok..
```
q3 524: sh bld
+ c++ -DA t.cc
+ c++ -DB t.cc
+ c++ -DC t.cc
+ c++ -DD t.cc
+ c++ -DE t.cc
q3 525: cat t.cc

#include <iostream>
using namespace std;

int main()
{
#ifdef A
int s{200};
#endif
#ifdef B
int s(200);
#endif
#ifdef C
int s = (200);
#endif
#ifdef D
int s = {200};
#endif
#ifdef E
int s = 200;
#endif
cout << s << endl;
}

```
See my earlier comments for a weak explanation as to why
this might fail on Macos.
Hope this helps

**Reply by yyueyangg** [reply]
yep, I have seen all your comments and replies. It sure helps to know that I am not the only one who have experienced this problem. thank you and I appreciate your help. but may I clarify further with u that, in ur earlier comments, ur explanation stated that this chosen form of initialisation was made legal in c++11, so does that mean that if I were to switch my version of c++ to 11 it will work? because I tried doing that and it doesn't. or does this form of initialisation just not work for Mac OS at all?

**Reply by GhostOf0days** [reply]
Programming languages don't differ across operating systems.

**Comment by davea42** [comment]
GhostOf 0days is correct, but languages do differ across language versions.
And the warnings/error reports from compilers do differ in important ways (see below).

Under linux, with the -std=c++08  and -DA options and GNU C++
(the -std is the one of real  interest) the compile fails.
Essentially this shows my earlier reference to the C++ standard version
and aggregate initializers was appropriate.

I should have tried that earlier (sorry).

I think an appropriate general rule is to restrain the impulse to believe
that every example of what to use in a given language is  necessarily
a good idea.  To become accomplished one should aim to use the parts
of the language that are well known and not become fixated on
some particular feature.  You don't want to find your software simply
won't work on an earlier compiler or some other compiler.

If some feature is new in, say C++20, you really
do NOT want to use it.   Using features just because they are
documented is not  professional behavior.  Unless there
is some overriding reason in a particular situation,
and then you really must document the feature  use in the code.

Stick with  C++98 unless you really need later features.
Code you find on the web is unlikely to indicate what
version of the language is required!

Another thing to note is that finding access to multiple environments
(Macos, Windows/Mingw64,Linux) to test in is an education since
compilers differ in what they care to warn about.  There is little
need to find access to other environments too soon. But
before you think you are done you will find it rewarding to
compile with other compilers.

```
+ c++ -std=c++98 -DA t.cc
t.cc: In function ‘int main()’:
t.cc:8:6: warning: extended initializer lists only available with ‘-std=c++11’ or ‘-std=gnu++11’
    8 | int s{200};
      |      ^

```
The other -D options do not get an error  or warning with -std=c++98
I hope this is clearer now.

**Reply by yyueyangg** [reply]
Wow! That is an ideal perspective from u to see the big picture and not be fixated on some particular features of a language, but rather to aim to apply the other parts that are more well known. It is also very helpful to understand the need for my codes / software to be able to compile not just on my device and system, but on others' as well. I haven't tried compiling with other compilers or even test it in other environments before. I guess I should definitely be learning from u. Currently, I am serving my national service in sg and only get to code whenever I am free. I will try to improve my skills by learning knowledges on compilers and testing environment other than really just practicing on programming languages itself

**Comment by davea42** [comment]
On Macos 

c++ -DA -std=c++11  t.cc
compiles without a warning.
Sorry, I should have tried this earlier.

My compiler version on Macos is newer than yours
(the machine is on  Macos Ventura 13.3.1).
So the -std may not work for you.

**Comment by davea42** [comment]
I think it is safe to assume that the book you reference says explicitly (possibly in a Forword section
or an Introduction) what version of C++ they are assuming.

I suggest you look for that information, it will help you with other examples from the book.

Evidently the 'properties' settings you are using are not working.
Are you using the command line or an IDE?

**Reply by yyueyangg** [reply]
I was looking at C++ CRASH COURSE by Josh Lospinoso published by No Starch Press. The copy of the book can be downloaded as a pdf copy from the internet. The forword section didn't explicitly specify what version of c++ will be taught, but rather refer to them as a whole with "Modern C++". 

I am using Visual Studio code as my IDE rather than the usage of code::block suggested by the book. From the interface in VS code, the bottom left icon allows us to access the settings by clicking on the settings option. that's where I clicked in to change the Cpp Standard to other versions. But it seems like I have never been able to change it to c++11 standard. Some of the codes that I have learnt from c++ give me a warning that the actions are deprecated as it is part of the c++11 standard, proving I have never been able to change my version correctly. Do u know how I can fix this and change my version of c++ the right way? 

Additionally, I would like to thank you for helping me out, your perspective as an experienced coder truly helps for a learner like me especially when I am self learning coding skills unguided by anyone while serving the army

**Reply by ChilledFerrum9** [reply]
@yyueyangg were you able to find out a solution for this? Really need help with this.

**Comment by davea42** [comment]
I've been coding a long time and never warmed to the idea of using an IDE.
I just use the command line for coding.
But that's just me.

So, no, sorry, I have no idea at all how to change the version in your IDE.

My Mac machine is on Clang 14.0.3 whereas you mention clang 12.0.0
According to` clang.llvm.org/cxx_status.html`
C++11 was supported way back in clang 3.3, so I'm mystified as to
why you've been unable to select c++11.
